Src/Common/FSecure/Mattermost/MattermostApi.cpp

```diff
@@ -0,0 +1,636 @@
+#include "stdafx.h"
+#include "MattermostApi.h"
+#include "Common/FSecure/CppTools/StringConversions.h"
+#include "Common/FSecure/WinHttp/HttpClient.h"
+#include "Common/FSecure/WinHttp/Constants.h"
+#include "Common/FSecure/WinHttp/Uri.h"
+#include <random>
+#include <set>
+#include <cctype>
+#include <algorithm>
+#include <sstream>
+
+using namespace FSecure::StringConversions;
+using namespace FSecure::WinHttp;
+
+namespace
+{
+	std::wstring ToWideString(std::string const& str)
+	{
+		return Convert<Utf16>(str);
+	}
+}
+
+FSecure::Mattermost::Mattermost(std::string const& serverUrl, std::string const& userName, std::string const& teamName, std::string const& accessToken, std::string const& channelName, std::string const& userAgent)
+{
+	if (auto winProxy = WinTools::GetProxyConfiguration(); !winProxy.empty())
+		this->m_ProxyConfig = (winProxy == OBF(L"auto")) ? WebProxy(WebProxy::Mode::UseAutoDiscovery) : WebProxy(winProxy);
+
+    this->m_ServerUrl = serverUrl;
+	this->m_AccessToken = accessToken;
+	this->m_UserAgent = userAgent;
+	this->m_OriginalChannelName = channelName;
+	this->m_UserName = userName;
+
+	std::string lowerChannelName = channelName;
+	std::transform(lowerChannelName.begin(), lowerChannelName.end(), lowerChannelName.begin(), [](unsigned char c) { return std::tolower(c); });
+
+    SetTeamID(FindTeamID(teamName));
+	SetChannel(CreateChannel(lowerChannelName));
+}
+
+void FSecure::Mattermost::SetTeamID(std::pair<std::string, std::string> const& teamID)
+{
+    this->m_TeamID = teamID.first;
+	this->m_TeamName = teamID.second;
+}
+
+void FSecure::Mattermost::SetUserAgent(std::string const& userAgent)
+{
+	this->m_UserAgent = userAgent;
+}
+
+void FSecure::Mattermost::SetChannel(std::string const& channelId)
+{
+	this->m_ChannelID = channelId;
+}
+
+void FSecure::Mattermost::SetToken(std::string const& accessToken)
+{
+    this->m_AccessToken = accessToken;
+}
+
+std::string FSecure::Mattermost::WritePost(std::string const& message, std::string const& fileID /* = "" */)
+{
+	return WritePostOrReply(message, fileID);
+}
+
+std::string FSecure::Mattermost::WriteReply(std::string const& message, std::string const& postID, std::string const& fileID /* = "" */)
+{
+    return WritePostOrReply(message, postID, fileID);
+}
+
+std::string FSecure::Mattermost::GetUserId(std::string const& userName)
+{
+	if (userName.empty())
+		return "";
+
+	std::string url = m_ServerUrl + OBF("/api/v4/users/") + userName;
+	auto response = GetJsonResponse(url);
+
+	if (response.value(OBF("status_code"), 0) == 400 || !response.contains(OBF("roles")))
+	{
+		url = m_ServerUrl + OBF("/api/v4/users/username/") + userName;
+		response = GetJsonResponse(url);
+
+		if (response.value(OBF("status_code"), 0) == 400 || !response.contains(OBF("roles")))
+		{
+			throw std::exception(OBF("[x] Could not find user specified by neither ID nor username!\n"));
```

Comment:
```suggestion
			throw std::runtime_error ...
```

std::exception is base class of family of exceptions. string constructor is non standard, and MSVC should not allow it.